arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Bracket.kt

```diff
@@ -89,16 +87,19 @@ public suspend inline fun <A> guarantee(
 public suspend inline fun <A> guaranteeCase(
   fa: suspend () -> A,
   crossinline finalizer: suspend (ExitCase) -> Unit
-): A {
-  val res = try {
-    fa()
-  } catch (e: CancellationException) {
-    runReleaseAndRethrow(e) { finalizer(ExitCase.Cancelled(e)) }
-  } catch (t: Throwable) {
-    runReleaseAndRethrow(t.nonFatalOrThrow()) { finalizer(ExitCase.Failure(t)) }
+): A = finalizeCase({ fa() }) { ex ->
+  try {
+    withContext(NonCancellable) {
+      finalizer(ex)
+    }
+  } catch (e: Throwable) {
+    e.nonFatalOrThrow()
+    when (ex) {
```

Comment:
This can be written as `(ex.errorOrNull ?: throw e).addSuppressed(e)` but I went for clarity instead, so that the code now reads like "if fa completed successfully, rethrow the error from its finalizer. Otherwise, suppress that error"